src/service/cid-impl.js

```diff
@@ -383,52 +390,122 @@ function setCidCookie(win, scope, cookie) {
 }
 
 /**
- * If cookie exists it's returned immediately. Otherwise, if instructed, the
- * new cookie is created.
- *
+ * Sets a new CID backup in Storage
+ * @param {!./ampdoc-impl.AmpDoc} ampdoc
+ * @param {string} cookieName
+ * @param {string} cookie
+ */
+function setCidBackup(ampdoc, cookieName, cookie) {
+  Services.storageForDoc(ampdoc).then((storage) => {
+    const isViewerStorage = storage.isViewerStorage();
+    if (!isViewerStorage) {
+      const key = getStorageKey(cookieName);
+      storage.setNonBoolean(key, cookie);
+    }
+  });
+}
+
+/**
+ * @param {string} cookieName
+ * @return {string}
+ */
+function getStorageKey(cookieName) {
+  return 'amp-cid:' + cookieName;
+}
+
+/**
+ * Maybe gets the CID from cookie or, if allowed, gets backup CID
+ * from Storage and checks validitiy.
  * @param {!Cid} cid
  * @param {!GetCidDef} getCidStruct
- * @param {!Promise} persistenceConsent
- * @return {!Promise<?string>}
+ * @return {!Promise<?string|undefined>}
```

Comment:
Personally I prefer `null` and avoid `undefined` as much as we can.
```
{if (!backupCid) {
  return null
}
setCidCookie()
setCidBackup();
return backupCid
}
return Promise.resolve(null)
```

It's personally preference, choose which ever you prefer. But I believe it should be either
`{!Promise<?string>}` or `{!Promise<string|undefined}`?